doc/sphinx/proof-engine/vernacular-commands.rst

```diff
@@ -571,25 +562,31 @@ Chapter :ref:`thecoqcommands` for documentation on how to compile a file). A com
 file is a particular case of module called *library file*.
 
 
-.. cmd:: Require @qualid
+.. cmd:: Require {? {| Import | Export } } {+ @qualid }
 
-   This command looks in the loadpath for a file containing module :n:`@qualid`
-   and adds the corresponding module to the environment of |Coq|. As
-   library files have dependencies in other library files, the command
-   :cmd:`Require` :n:`@qualid` recursively requires all library files the module
-   qualid depends on and adds the corresponding modules to the
-   environment of |Coq| too. |Coq| assumes that the compiled files have been
-   produced by a valid |Coq| compiler and their contents are then not
-   replayed nor rechecked.
+   Loads compiled modules into the |Coq| environment.  The command
+   searches the loadpath for the modules named by the :n:`@qualid`\s.  Each :n:`@qualid`
+   has the form :n:`@dirpath.@ident`.  The commmand maps :n:`@dirpath` to a physical
+   directory (see Section :ref:`libraries-and-filesystem`) and loads the compiled
+   file :n:`@ident.vo` from that directory.  (Compiling :n:`@ident.v` generates :n:`@ident.vo`.)
 
-   To locate the file in the file system, :n:`@qualid` is decomposed under the
-   form :n:`dirpath.@ident` and the file :n:`@ident.vo` is searched in the physical
-   directory of the file system that is mapped in |Coq| loadpath to the
-   logical path dirpath (see Section :ref:`libraries-and-filesystem`). The mapping between
+   The process is applied recursively to all the loaded modules.
+   The compiled files must have been compiled with the same version of |Coq|.
+   The compiled files are neither replayed nor rechecked.
+
+   * :n:`Import` - TBA
+   * :n:`Export` - TBA
+
+   .. todo: I can't make sense of the old explanations for Import and Export
+
+   The mapping between
    physical directories and logical names at the time of requiring the
    file must be consistent with the mapping used to compile the file. If
    several files match, one of them is picked in an unspecified fashion.
 
+   .. todo both sentences still true?  Second sentence is unsettling
```

Comment:
Yes. Usually when something is documented as "unspecified", it is a way of warning the users not to rely on it. That being said there are many other things in Coq that are unspecified and that users rely on.

Reply:
I would have expected the mapping would be deterministic, like searching for a file in a list of paths like searching for an executable as in Linux or Windows.  And the consistency requirement sounds more like a bug than a feature.

"consistent" deserves some explanation--you and I may install Coq in different directories but the .vos in the distribution are still consistent.

I could just leave a todo on the definition of "consistency" here.

Reply:
Unspecified isn't the same as non-deterministic. It could be the case that it is filesystem-dependent for instance.

Regarding the definition of consistent, this is a standard error that users do encounter and would deserve being properly documented (with `.. exn::`). Here's an example:
```
echo "Definition foo := 0." > foo.v
coqc -R . my_project foo.v
echo "Require Import foo." > bar.v
coqc bar.v
(*
Error: The file /tmp/foo.vo contains library my_project.foo and not library foo
*)
```

Usually, this error occurs not with `coqc` (there is no reason why the user would use `-R` on the first invocation but not on the second one) but with `coqide` (doing `coqide bar.v` and trying to step through it). The typical case where this occurs is the following:

```
cat > Make <<< '-R . my_project
foo.v
bar.v'
coq_makefile -f Make -o Makefile
make
coqide bar.v
```

A learner downloads a project with a Makefile generated from a `Make` file containing the `-R` option. When they try to step through the file with an IDE, they encounter the error mentioned above. The trick is either to run `coqide -R . my_project bar.v` or to rename `Make` into `_CoqProject` because then the IDE will read the first line of `_CoqProject` to know the option to use.

If you have time to translate this explanation into a clear description of the error message (ideally with an example) in the manual, that would be really useful!

Reply:
Ah, what is the definition of "consistent"?  A bit hard to really understand the error you describe without knowing what's correct.

The command line options for CoqIDE are not documented.  I didn't know you could do `coqide bar.v`.  However, when I try `coqide --help` (8.10.1 in Windows), the GUI doesn't come up and there's no message on stdout.

The default `Top` is not documented, nor are command line options for coq_makefile, not even mentioning that `coq_makefile --help` lists the parameters.  This line in the --help output seems out of date:  `[file.ml[i4]?]: Objective Caml file to be compiled`

I don't see any clear definition of `loadpath`.  Some of the text in gallina-extensions touches on it but doesn't define it.

BTW, I notice that `:-Q *directory* *dirpath*:` in `coq-commands.rst` drops the space between the 2 items:

![image](https://user-images.githubusercontent.com/1253341/78420517-ad708280-7604-11ea-8f30-a3cd2f7524d5.png)

Seems like we should take this up a bit later.  I will leave a todo linking to your comment.

> Unspecified isn't the same as non-deterministic. It could be the case that it is filesystem-dependent for instance.

This still sounds completely bogus.   Perhaps we should discuss outside this PR, but...

From the doc "If several files with identical base name are present in different subdirectories of a recursive loadpath, which of these files is found first may be system- dependent and explicit qualification is recommended."  Either a) there should be an error message when you try to import the ambiguous item or b) the order of items in the loadpath should control which one is picked when it's ambiguous.

I don't see any reason why this should be filesystem-dependent.  AFAIK Coq is not relying on the filesystem to interpret the loadpath.  If someone adds an additional item to the loadpath (e.g. for a new external library), it seems like this would cause a bunch of immediate problems.

Reply:
> Ah, what is the definition of "consistent"? A bit hard to really understand the error you describe without knowing what's correct.

They must use the same `-R` of `-Q` prefix.

> The command line options for CoqIDE are not documented.

That's too bad (and worth an issue)! Most of the `coqtop` / `coqc` options are accepted in fact.



> The default `Top` is not documented, nor are command line options for coq_makefile, not even mentioning that `coq_makefile --help` lists the parameters. This line in the --help output seems out of date: `[file.ml[i4]?]: Objective Caml file to be compiled`

Worth another issue! But the most important ones are documented. In fact the only command-line option that a user should give to `coq_makefile` are `-f` (usually `-f _CoqProject`) and `-o` (usually `-o Makefile` or `-o Makefile.coq` if there is a small wrapper `Makefile`). The other options should be passed as the first line of `_CoqProject` (and those that are accepted are documented). I think these are the same that are accepted directly on the command-line, but this is not recommended.

>This still sounds completely bogus.

That's the way it is and do not count on me to fix this. As long as it is like that, it is important to document it to warn users. Users should always prefer the `From ... Require` form that can guarantee from which library they are loading a module. And library authors should ensure that no two modules of their library are named the same. With these two good practices, you are safe from this problem.

Reply:
The error "The file /tmp/foo.vo contains library my_project.foo and not library foo" is already listed under From ... Require

> That's the way it is and do not count on me to fix this.

Not asking for that at all, for now just trying to understand whether this is indeed a weak point in Coq or the doc and if there's a solid reason for it.  We don't need to resolve this now.

I took a crack at adding some of the information you suggested in the `Libraries and filesystem` section of Gallina extensions.  Maybe I should just back that out for now--seems like too much to tack on to this PR.  The documentation for this feature is scattered/repeated in several different sections.  That makes it easy to get confused.  Also there seem to be 4 or 5 different ways to set up logical names, including by resource file and environment variables.  Maybe that's too many, or we should at least pick one or two and recommend them as the preferred directory.  But the underlying mechanism seems (probably) simple.

And how is `_CoqProject` found?  What directory(ies) are examined?

Reply:
>  The documentation for this feature is scattered/repeated in several different sections. That makes it easy to get confused.

Yes, and we will need to consolidate it in a single place (cf. #12012).



> there seem to be 4 or 5 different ways to set up logical names, including by resource file and environment variables. Maybe that's too many, or we should at least pick one or two and recommend them

Agreed, and in fact I've already a clear view of what are the recommended best practices and what practices are discouraged, but it's a bit hard to share it like this, out of specific examples.

> And how is `_CoqProject` found? What directory(ies) are examined?

I guess you mean by IDEs (since `coq_makefile` won't "find" `_CoqProject`, you need to give it on the command-line). IDEs that I know will look for `_CoqProject` in the directory you open. E.g., if you run `coqide &` from the command-line in a specific directory, or `code . &` for VsCode (or equivalently the menu "Open directory").